azurelinuxagent/common/utils/restutil.py

```diff
@@ -98,43 +146,63 @@ def _http_request(method, host, rel_uri, port=None, data=None, secure=False,
     return resp
 
 
-def http_request(method, url, data, headers=None, max_retry=3,
-                 chk_proxy=False):
-    """
-    Sending http request to server
-    On error, sleep 10 and retry max_retry times.
-    """
+def http_request(method,
+                url, data, headers=None,
+                chk_proxy=False,
+                max_retry=DEFAULT_RETRIES,
+                retry_codes=RETRY_CODES,
+                retry_delay=SHORT_DELAY_IN_SECONDS):
+
+    global SECURE_WARNING_EMITTED
+
     host, port, secure, rel_uri = _parse_url(url)
-    global secure_warning
 
     # Check proxy
     proxy_host, proxy_port = (None, None)
     if chk_proxy:
-        proxy_host, proxy_port = get_http_proxy()
+        proxy_host, proxy_port = _get_http_proxy()
 
-    # If httplib module is not built with ssl support. Fallback to http
+    # If httplib module is not built with ssl support,
+    # fallback to HTTP if allowed
     if secure and not hasattr(httpclient, "HTTPSConnection"):
+        if not conf.get_allow_http():
+            raise HttpError("HTTPS is unavailable and required")
```

Comment:
Ok, so we will just keep looping? I thought perhaps we might want to cleanly exit the process if there is no hope of continuing on.